fix(core): 18465 tune redirection, react router migrated to v6

vkozio opened the pull request:
https://kontur.fibery.io/Tasks/Task/Tune-redirection-on-Atlas-app-18465

- Routing overhaul

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Enhanced routing setup with new components and improved logic for better navigation experience in the application.
  - Added new environment variables for testing and Slack bot integration.
  - Introduced new localization string for error handling related to an incorrect MCDA version.
  - Added support for uploading test reports to surge.sh and workflow artifacts.

- **Bug Fixes**
  - Improved error handling in MCDA layer parameters to safely manage potential `null` or `undefined` values.

- **Refactor**
  - Updated dependency versions for better compatibility and security.
  - Refined MCDA configuration handling by focusing on the `name` property instead of `id`, ensuring consistent and clear configurations.
  
- **Chores**
  - Scheduled automated e2e tests every two hours on weekdays.
  - Added detailed Slack reporter configurations for test notifications.
  - Removed unused exports of `Column` and `Row` components from `Layout`.

- **Tests**
  - Added new environment variables and configurations for e2e tests, including Keycloak admin credentials and Slack bot token.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Commit:
router overhaul

Commit:
event-based navigation

Commit:
use map feature

Commit:
rebase errors fix

Komzpa commented on src/core/router/types.ts:
```diff
@@ -21,6 +21,8 @@ export interface AppRoute {
    * */
   visibilityInNavigation?: 'always' | 'never' | 'auto';
+  /** Visible but disabled in sidebar */
+  disabled?: boolean;
```
@coderabbitai generate first documentation example if you suggest adding it

albaranau committed:
fix(18774): Fix crash when opening MCDA file without indicators array (#763)

albaranau committed:
fix(18079): Add name parameter to MCDA configuration. Refactor opening mcda json file (#764)

* fix(18079): add name parameter to MCDA config

* fix(18079): handle MCDA files with missin id/name. Refactor MCDA upload

amurKontur committed:
[FE+BE] Automate simple registration on dev and test (#747)

* test: adapt workflow and go to keycloak for registration test

* test: add registration, verification of email and login by registered user actions to registration test

* test: update tests and improve workflow run name

* test: added delete user action to registration test

* test: review fixes and other tests fixes

* test: fix typo

* test: fix variables names

Commit:
refactor initial redirect

Commit:
misc fix

Commit:
streamline landing detect

dqunbp commented on src/core/auth/atoms/userWasLanded.ts:
```diff
@@ -1,25 +1,11 @@
+  const noSettingUrl =
+    location.pathname === configRepo.get().baseUrl && initialUrl.search === '';
+  const userHaveLandedMark = localStorage.getItem('landed');
+  return !(noSettingUrl && !userHaveLandedMark);
```
This is the same as `return !noSettingUrl || userHaveLandedMark;`

dqunbp commented on src/core/auth/atoms/userWasLanded.ts:
```diff
@@ -1,25 +1,11 @@
 export const landUser = () => localStorage.setItem('landed', 'true');
+export function userWasLanded() {
+  const initialUrl = new URL(localStorage.getItem('initialUrl') || '');
+  const noSettingUrl =
```
Using negation in variable names complicates code comprehension. Better to avoid it

dqunbp commented on src/core/auth/atoms/userWasLanded.ts:
```diff
@@ -1,25 +1,11 @@
-
 export const landUser = () => localStorage.setItem('landed', 'true');
+export function userWasLanded() {
+  const initialUrl = new URL(localStorage.getItem('initialUrl') || '');
```
We will get `TypeError: Failed to construct 'URL': Invalid URL` if LS does not contain initialUrl.

Commit:
review fixes

Commit:
fix CI

albaranau committed:
fix(18607): MCDA style: paint faulty values (like null and NaN) as transparent (#766)

albaranau committed:
fix(mcdaStyle): 18787 fix style calculation for MCDA with a single axis (#767)

* fix(mcdaStyle): 18787 fix style calculation for MCDA with a single axis

* Revert changes that duplicate different PR

Commit:
fix names

amurKontur committed:
Add a flow to run tests every 2 hours, report results to slack (#765)

* test: slack integration and workflow update

* test: fix workflow

* test: fix paths

* test: use npm to install

* test: clear cache to install

* test: add slack/web-api

* test: adding html report

* test: add clean up for github pages

* test: use surge.sh instead of pages

albaranau committed:
Version 2.29.1 (#768)

albaranau commented on src/views/CommonView.tsx:
```diff
@@ -2,11 +2,11 @@ import { Suspense, useEffect } from 'react';
 import { lazily } from 'react-lazily';
 import { useAtom } from '@reatom/react-v2';
 import { configRepo } from '~core/config';
-import { Row } from '~components/Layout';
```
Was there an intention to remove `Row` component?
Because it isn't used anywhere else, maybe we should remove it (same about `Column`)

vkozio replied:
done

albaranau commented on src/core/router/components/Router.tsx:
```diff
@@ -1,26 +1,106 @@
+  createBrowserRouter,
+  type RouteObject,
+} from 'react-router-dom';
+import { KeepAliveProvider } from 'react-component-keepalive-ts';
```
Are we sure we want to switch to this library? It looks very obscure and less maintained than `react-router-cache-route`

https://www.npmjs.com/package/react-component-keepalive-ts
<img width="430" alt="image" src="https://github.com/konturio/disaster-ninja-fe/assets/9570735/808039b9-711a-44ff-942b-1ab0997f8e5a">

vkozio replied:
They all are "less maintained" actually, when you look into commits. Just switched to lib, compatible with current deps

albaranau commented on src/core/apiClientInstance.ts:
```diff
@@ -10,8 +9,10 @@ export const apiClient = withSetupCheck(ApiClient, {
           alert('Access denied or your session expired.');
-          goTo('/profile');
-          location.reload();
+          import('~core/router/goTo').then(({ goTo }) => {
```
Is it required to import this dynamically?

albaranau commented on src/core/auth/atoms/userWasLanded.ts:
```diff
@@ -1,25 +1,16 @@
+  const initialUrlObj = new URL(initialUrl);
+  const isHomepageOnFirstVisit =
+    location.pathname === configRepo.get().baseUrl && initialUrlObj.search === '';
+  return !isHomepageOnFirstVisit;
```
I have a bit of a hard time understanding all of the logic here. Maybe we could have a catch up?

albaranau commented on src/core/router/getAbsoluteRoute.ts:
```diff
@@ -15,7 +15,7 @@ const pathJoin = (...path: Array<string>) => {
 export const getAbsoluteRoute = (
   slugOrRoute: string | AppRoute,
-  base = configRepo.get().baseUrl,
+  base = '', //configRepo.get().baseUrl,
```
is it supposed to be commented?

albaranau commented:
I tried out the PR, and there's this new flickering of the Sidebar (see video):
https://github.com/konturio/disaster-ninja-fe/assets/9570735/6786cb52-1b6a-4a8e-8905-b3e228774134

albaranau commented on src/features/user_profile/components/SettingsForm/SettingsForm.tsx:
```diff
@@ -205,29 +186,24 @@ function SettingsFormGen({ userProfile, updateUserProfile }) {
-                  <Select
-                    alwaysShowPlaceholder
+                {(featureFlags[FeatureFlag.FEED_SELECTOR] ||
+                  featureFlags[FeatureFlag.EVENTS_LIST__FEED_SELECTOR]) && (
```
What's the difference between the two flags? shouldn't one of them be deprecated?

vkozio replied:
to refactor in next iteration of feature configs

albaranau commented on src/core/shared_state/currentEventFeed.ts:
Are the Event Feed changes a part of some other ticket or is it an unrelated refactoring?

albaranau commented on src/core/url_store/atoms/urlStore.ts:
```diff
@@ -36,10 +36,9 @@ export const urlStoreAtom = createAtom(
-      FeatureFlag.FEED_SELECTOR,
-    ].some((flag) => typeof configRepo.get().features[flag] !== 'undefined');
+    const isFeedSelectorEnabled =
+      configRepo.get().features['events_list__feed_selector'] ||
```
Why use bare strings instead of constants (FeatureFlag.EVENTS_LIST__FEED_SELECTOR and FeatureFlag.FEED_SELECTOR) here?

vkozio replied:
to be refactored during config simplification

albaranau commented:
Would be nice to rename the PR to include the most important changes it brings (react-router dom update, routes refactoring, changes in redirects) so it's obvious from the squashed commit name

albaranau commented on src/features/user_profile/components/SettingsForm/SelectFeeds.tsx:
Shouldn't this component be in /features?

vkozio replied:
no, it's used only in profile

dqunbp committed:
feat: Add public access to refreshAuthToken method (#771)

The `refreshAuthToken` method in the `ApiClient` class was changed from private to public. This change allows external code to refresh the authentication token. This is necessary to get fresh roles data from Keycloak.

albaranau commented on src/core/auth/atoms/userWasLanded.ts:
```diff
@@ -1,25 +1,16 @@
+  if (initialUrl === null) return true;
+  const initialUrlObj = new URL(initialUrl);
+  const isHomepageOnFirstVisit =
+    location.pathname === configRepo.get().baseUrl && initialUrlObj.search === '';
```
Can we maybe add a comment here?

Commit:
deps fixes and updates after merge

Commit:
review fixes

Commit:
deps update

vkozio committed:
Merge branch 'main' into 18465-tune-redirection-on-atlas-app

albaranau commented on package.json:
Approving, but let's add smth like "React Router updated to v6" to PR name, because it's good to keep this info visible in commit history
